### PR TITLE
[Coop] Convert ves_icall_System_Array_FastCopy.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -774,19 +769,25 @@ ves_icall_System_Array_FastCopy (MonoArray *source, int source_idx, MonoArray* d
 			return FALSE;
 	}
 
+	gchandle_t source_gchandle = 0;
```
Comment (jaykrell) [inline mono/metadata/icall.c]: Might want to avoid the pinning.
Or find a way to use `fixed`.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/metadata/icall.c [hunk 2]
```diff
@@ -774,61 +761,66 @@ ves_icall_System_Array_FastCopy (MonoArray *source, int source_idx, MonoArray* d
...
 	}
 
+	gboolean pin = FALSE; // FIXME Making a gchandle is inefficient and breaks profiler tests.
+	gchandle_t source_gchandle = 0;
+	gchandle_t dest_gchandle = 0;
+
 	if (m_class_is_valuetype (dest_class)) {
-		element_size = mono_array_element_size (source->obj.vtable->klass);
-		source_addr = mono_array_addr_with_size_fast (source, element_size, source_idx);
+		gsize const element_size = mono_array_element_size (MONO_HANDLE_GETVAL (source, obj.vtable->klass));
```
Comment (lambdageek) [inline mono/metadata/icall.c]: I think this block is another great place to use `MONO_ENTER_NO_SAFEPOINTS`.
And to delete `gboolean pin`.

Comment (jaykrell) [reply mono/metadata/icall.c]: I have no other known options, so yeah. Maybe could be builtin to the wrapper to reduce TLS access.

Comment (jaykrell) [reply mono/metadata/icall.c]: Would also be nice for icall wrapper to pass down the `ThreadInfo`.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build  OS X x64 C++

**Comment by jaykrell** [comment]
@monojenkins build Windows x64

**Comment by jaykrell** [comment]
@monojenkins build OS X  i386

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 C++

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Pull request merged**